In [1]:
import torch
print(torch.__version__)

2.11.0+cu126


## Tensors

In [2]:
# Scalars
scalar = torch.tensor(7)
scalar

tensor(7)

In [3]:
# Get tensor as Python int
scalar.item()

7

In [4]:
scalar.ndim

0

In [5]:
# Vectors
vector = torch.tensor([7, 7])
vector

tensor([7, 7])

In [6]:
vector.ndim

1

In [7]:
vector.shape

torch.Size([2])

In [8]:
# Matrix
matrix = torch.tensor([[7, 8],
                       [9, 10]])
matrix

tensor([[ 7,  8],
        [ 9, 10]])

In [9]:
matrix.ndim

2

In [10]:
matrix.shape

torch.Size([2, 2])

In [11]:
# Tensor
tensor = torch.tensor([[[1, 2], [3, 4]],
                       [[5, 6], [7, 8]]])
tensor

tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])

In [12]:
tensor.ndim

3

In [13]:
tensor.shape

torch.Size([2, 2, 2])

In [14]:
tensor[0]

tensor([[1, 2],
        [3, 4]])

In [15]:
tensor[0][0]

tensor([1, 2])

In [16]:
tensor[1]

tensor([[5, 6],
        [7, 8]])

### Random Tensors

In [17]:
# Random tensor of size (3, 4)
rand_tensor = torch.rand(3, 4)
rand_tensor

tensor([[0.5513, 0.2805, 0.9756, 0.8186],
        [0.1264, 0.4798, 0.7162, 0.1968],
        [0.9560, 0.5917, 0.8361, 0.4074]])

In [18]:
# Random tensor with similar size to image tensor
random_image_tensor = torch.rand(size=(224, 224, 3)) # height, width, color-channel(R, G, B)
random_image_tensor.shape, random_image_tensor.ndim

(torch.Size([224, 224, 3]), 3)

### Zeros and Ones

In [19]:
# Tensor with all zeros
zeros = torch.zeros(size=(3, 4))
zeros

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [20]:
# Tensor with all ones
ones = torch.ones(size=(3, 4))
ones

tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])

### Range of tensors and tensors-like

In [21]:
one_to_nine = torch.arange(1, 10)
one_to_nine

tensor([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [22]:
# Tensor like
ten_zeros = torch.zeros_like(input=one_to_nine)
ten_zeros

tensor([0, 0, 0, 0, 0, 0, 0, 0, 0])

### Datatypes

In [23]:
float_32_tensor = torch.tensor([3.0, 6.0, 9.0],
                               dtype=None,
                               device=None,
                               requires_grad=False)
float_32_tensor

tensor([3., 6., 9.])

In [24]:
float_32_tensor.dtype

torch.float32

In [25]:
float_16_tensor = float_32_tensor.type(torch.float16)
float_16_tensor, float_16_tensor.dtype

(tensor([3., 6., 9.], dtype=torch.float16), torch.float16)

In [26]:
float_16_tensor * float_32_tensor

tensor([ 9., 36., 81.])

### Details about the tensor

In [27]:
# Create a tensor
random_tensor = torch.rand(size=(3, 4))
random_tensor

tensor([[0.7099, 0.9048, 0.7587, 0.8223],
        [0.2823, 0.7376, 0.5894, 0.5252],
        [0.9040, 0.4020, 0.6245, 0.2902]])

In [28]:
print(random_tensor)
print(f"Datatype of tensor: {random_tensor.dtype}")
print(f"Shape of tensor: {random_tensor.shape}")
print(f"Device tensor is on: {random_tensor.device}")

tensor([[0.7099, 0.9048, 0.7587, 0.8223],
        [0.2823, 0.7376, 0.5894, 0.5252],
        [0.9040, 0.4020, 0.6245, 0.2902]])
Datatype of tensor: torch.float32
Shape of tensor: torch.Size([3, 4])
Device tensor is on: cpu


### Manipulating tensors

Tensor operations includes:
* Addition
* Subtraction
* Multiplication (element-wise)
* Division
* Matrix multiplication

In [29]:
# Create a tensor
arith_tensor = torch.tensor([1, 2, 3])
arith_tensor

tensor([1, 2, 3])

In [30]:
arith_tensor + 10

tensor([11, 12, 13])

In [31]:
arith_tensor - 1

tensor([0, 1, 2])

In [32]:
# Elementwise multiplication
arith_tensor.mul(10)

tensor([10, 20, 30])

In [33]:
torch.mul(arith_tensor, 10)

tensor([10, 20, 30])

In [34]:
torch.add(arith_tensor, 10)

tensor([11, 12, 13])

Matrix multiplication

In [35]:
# Elementwise
print("Elementwise multiplication")
print(f"{arith_tensor} * {arith_tensor} = {torch.mul(arith_tensor, arith_tensor)}")

Elementwise multiplication
tensor([1, 2, 3]) * tensor([1, 2, 3]) = tensor([1, 4, 9])


In [36]:
# Matrix multiplication
print("Matrix multiplication")
print(f"Equals: {torch.matmul(arith_tensor, arith_tensor)}")

Matrix multiplication
Equals: 14


### Transpose

In [37]:
# transpose
mul_tensor = torch.tensor([[1, 2, 3],
                           [4, 5, 6]])
transposed_matrix = mul_tensor.mT
transposed_matrix, mul_tensor

(tensor([[1, 4],
         [2, 5],
         [3, 6]]),
 tensor([[1, 2, 3],
         [4, 5, 6]]))

In [38]:
mul_tensor.shape, transposed_matrix.shape

(torch.Size([2, 3]), torch.Size([3, 2]))

In [39]:
torch.mm(mul_tensor, transposed_matrix)

tensor([[14, 32],
        [32, 77]])

### Finding min, max, mean, sum, etc. (Tensor aggregation)

In [40]:
# Create a tensor
x = torch.arange(0, 100, 10)
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [41]:
# Find the min
torch.min(x), x.min()

(tensor(0), tensor(0))

In [42]:
# Find the max
torch.max(x), x.max()

(tensor(90), tensor(90))

In [43]:
# Find the avg (mean)
torch.mean(x.type(torch.float32)), x.type(torch.float32).mean()

(tensor(45.), tensor(45.))

In [44]:
# Find the sum
torch.sum(x), x.sum()

(tensor(450), tensor(450))

### Positional min and max

In [45]:
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [46]:
x.argmax()

tensor(9)

In [47]:
x.argmin()

tensor(0)

### Reshaping, stacking, squeezing and unsqueezing tensors

In [2]:
# Create tensor
import torch
x = torch.arange(1., 10.)
x, x.shape


(tensor([1., 2., 3., 4., 5., 6., 7., 8., 9.]), torch.Size([9]))

In [3]:
# Change the dimensions
x_reshaped = x.reshape(3, 3)
x_reshaped, x_reshaped.shape

(tensor([[1., 2., 3.],
         [4., 5., 6.],
         [7., 8., 9.]]),
 torch.Size([3, 3]))

In [4]:
# Change the view
z = x.view(1, 9)
z, z.shape

(tensor([[1., 2., 3., 4., 5., 6., 7., 8., 9.]]), torch.Size([1, 9]))

In [5]:
# Changing z changes x (because a view of a tensor shares same memory as the orginal tensor)
z[:, 0] = 5
z, x

(tensor([[5., 2., 3., 4., 5., 6., 7., 8., 9.]]),
 tensor([5., 2., 3., 4., 5., 6., 7., 8., 9.]))

In [6]:
# Stacking tensors on top of each other
x_stacked = torch.stack([x, x, x, x], dim=0)
x_stacked, x_stacked.shape

(tensor([[5., 2., 3., 4., 5., 6., 7., 8., 9.],
         [5., 2., 3., 4., 5., 6., 7., 8., 9.],
         [5., 2., 3., 4., 5., 6., 7., 8., 9.],
         [5., 2., 3., 4., 5., 6., 7., 8., 9.]]),
 torch.Size([4, 9]))

In [ ]:
# Squeezing -> torch.squeeze() - removes all single dimensions from a target tensor
x = torch.zeros(2, 1, 2, 1, 2)
x, x.shape

(tensor([[[[[0., 0.]],
 
           [[0., 0.]]]],
 
 
 
         [[[[0., 0.]],
 
           [[0., 0.]]]]]),
 torch.Size([2, 1, 2, 1, 2]))

In [15]:
x_squeezed = x.squeeze()
x_squeezed, x_squeezed.shape

(tensor([[[0., 0.],
          [0., 0.]],
 
         [[0., 0.],
          [0., 0.]]]),
 torch.Size([2, 2, 2]))

In [11]:
y = x.squeeze(0)
y, y.shape

(tensor([[[[[0., 0.]],
 
           [[0., 0.]]]],
 
 
 
         [[[[0., 0.]],
 
           [[0., 0.]]]]]),
 torch.Size([2, 1, 2, 1, 2]))

In [12]:
y = x.squeeze(1)
y, y.shape

(tensor([[[[0., 0.]],
 
          [[0., 0.]]],
 
 
         [[[0., 0.]],
 
          [[0., 0.]]]]),
 torch.Size([2, 2, 1, 2]))

In [13]:
y = x.squeeze(2)
y, y.shape

(tensor([[[[[0., 0.]],
 
           [[0., 0.]]]],
 
 
 
         [[[[0., 0.]],
 
           [[0., 0.]]]]]),
 torch.Size([2, 1, 2, 1, 2]))

In [14]:
y = x.squeeze(4)
y, y.shape

(tensor([[[[[0., 0.]],
 
           [[0., 0.]]]],
 
 
 
         [[[[0., 0.]],
 
           [[0., 0.]]]]]),
 torch.Size([2, 1, 2, 1, 2]))

In [19]:
# torch.unsqueeze() - adds a single dimension to a target tensor at a specific dim (dimension)
print(x_squeezed, x_squeezed.shape)
x_unsqueezed = x_squeezed.unsqueeze(dim=1)
print(x_unsqueezed, x_unsqueezed.shape)

tensor([[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]) torch.Size([2, 2, 2])
tensor([[[[0., 0.],
          [0., 0.]]],


        [[[0., 0.],
          [0., 0.]]]]) torch.Size([2, 1, 2, 2])


In [20]:
# torch.permute() - rearanges the dimensions of the tensor
x = torch.randn(2, 3, 5)
x.shape

torch.Size([2, 3, 5])

In [21]:
torch.permute(x, (2,0,1)).size()

torch.Size([5, 2, 3])

In [24]:
original_image = torch.rand(size=(244, 244, 3)) # [height, width, color-channel]

rearanged_image = original_image.permute(2, 0, 1).shape
rearanged_image# [color-channel, height, width]

torch.Size([3, 244, 244])

## Pytorch tensors and NumPy

NumPy is a popular scientific Python numerical computing library.

And because of this, PyTorch has functionality to interact with it.

* Data in NumPy, want in PyTorch tensor -> `torch. from_numpy(ndarray)`
* PyTorch tensor -> NumPy -> `torch.Tensor.numpy`

In [25]:
# NumPy array to PyTorch tensor
import torch
import numpy as np

array = np.arange(1.0, 8.0)
tensor = torch.from_numpy(array)
array, tensor

(array([1., 2., 3., 4., 5., 6., 7.]),
 tensor([1., 2., 3., 4., 5., 6., 7.], dtype=torch.float64))

In [26]:
# PyTorch tensor to NumPy array
tensor = torch.ones(7)
array = tensor.numpy()
tensor, array

(tensor([1., 1., 1., 1., 1., 1., 1.]),
 array([1., 1., 1., 1., 1., 1., 1.], dtype=float32))

### Reproducibility (trying to take random out of random)

In [1]:
import torch

# Create two random tensors
random_tensor_A = torch.rand(3, 4)
random_tensor_B = torch.rand(3, 4)

print(random_tensor_A)
print(random_tensor_B)
print(random_tensor_A == random_tensor_B)

tensor([[9.4534e-01, 2.3269e-01, 1.6598e-01, 8.1524e-01],
        [1.8553e-01, 4.8870e-04, 3.9438e-01, 3.0431e-01],
        [4.7204e-01, 8.9849e-01, 9.5545e-01, 3.3972e-01]])
tensor([[0.2324, 0.5568, 0.4159, 0.6480],
        [0.7710, 0.5218, 0.9059, 0.3027],
        [0.9285, 0.4709, 0.0668, 0.9576]])
tensor([[False, False, False, False],
        [False, False, False, False],
        [False, False, False, False]])


In [2]:
# Let's create some random but reproducible tensors

# Set a random seed
RANDOM_SEED = 42

torch.manual_seed(RANDOM_SEED)
random_tensor_C = torch.rand(3, 4)

torch.manual_seed(RANDOM_SEED)
random_tensor_D = torch.rand(3, 4)

print(random_tensor_C)
print(random_tensor_D)
print(random_tensor_C == random_tensor_D)

tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])
tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])
tensor([[True, True, True, True],
        [True, True, True, True],
        [True, True, True, True]])


### GPU setup

In [1]:
import torch
torch.cuda.is_available()

True

In [2]:
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

In [3]:
if torch.cuda.is_available():
    for i in range(torch.cuda.device_count()):
        props = torch.cuda.get_device_properties(i)
        print(f'GPU {i}: {props.name}')
        print(f'  Compute Capability: {props.major}.{props.minor}')
        print(f'  Total Memory: {props.total_memory / 1e9:.2f} GB')
else: print('CUDA not available')

GPU 0: NVIDIA GeForce RTX 3050 6GB Laptop GPU
  Compute Capability: 8.6
  Total Memory: 6.44 GB


In [4]:
# Create a tensor (default on CPU)
tensor = torch.tensor([1, 2, 3], device="cpu")

# Tensor not on GPU
print(tensor, tensor.device)

tensor([1, 2, 3]) cpu


In [5]:
# Move tensor on to the GPU (if availiable)
tensor_on_gpu = tensor.to(device)
tensor_on_gpu

tensor([1, 2, 3], device='cuda:0')

In [6]:
torch.cuda.device_count()

1

### Moving tensors to CPU

In [7]:
# If tensor is on GPU, can't transform it to NumPy
tensor_on_gpu.numpy()

TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

In [9]:
# To fix the GPU tensor with Numpy issue, we can first set it to the CPU
tensor_back_on_cpu = tensor_on_gpu.cpu().numpy()
tensor_back_on_cpu

array([1, 2, 3])

## Exercise (Solved)

2. Create a random tensor with shape `(7, 7)`.

In [1]:
# Q.2
import torch
random_tensor = torch.rand(7, 7)
random_tensor

tensor([[0.7788, 0.4692, 0.0457, 0.0176, 0.0760, 0.6351, 0.2417],
        [0.1425, 0.4649, 0.8156, 0.8871, 0.7369, 0.3109, 0.6002],
        [0.0709, 0.5001, 0.2833, 0.0207, 0.7866, 0.5497, 0.3542],
        [0.2082, 0.2639, 0.3255, 0.5261, 0.1308, 0.9753, 0.9427],
        [0.7603, 0.4547, 0.9234, 0.2367, 0.0407, 0.7820, 0.2825],
        [0.9690, 0.1785, 0.6250, 0.4646, 0.7512, 0.7592, 0.8189],
        [0.7048, 0.8808, 0.3240, 0.3423, 0.2967, 0.4733, 0.8299]])

3. Perform a matrix multiplication on the tensor from 2 with another random tensor with shape `(1, 7)` (hint: you may have to transpose the second tensor).

In [2]:
# Q.3
random_tensor_A = torch.rand(1, 7)
random_tensor_A.matmul(random_tensor)
random_tensor_A

tensor([[0.7304, 0.5471, 0.1757, 0.8723, 0.6366, 0.4266, 0.7030]])

4. Set the random seed to 0 and do 2 & 3 over again.
The output should be:

(tensor([[1.8542],
         [1.9611],
         [2.2884],
         [3.0481],
         [1.7067],
         [2.5290],
         [1.7989]]), torch.Size([7, 1]))

In [3]:
# Q.4
RANDOM_SEED = 0
torch.manual_seed(RANDOM_SEED)
tensor_with_seed = torch.rand(7, 1)
print(tensor_with_seed, tensor_with_seed.shape)

torch.manual_seed(RANDOM_SEED)
tensor_with_seed = torch.rand(7, 1)
print(tensor_with_seed, tensor_with_seed.shape)

torch.manual_seed(RANDOM_SEED)
tensor_with_seed = torch.rand(7, 1)
print(tensor_with_seed, tensor_with_seed.shape)

tensor([[0.4963],
        [0.7682],
        [0.0885],
        [0.1320],
        [0.3074],
        [0.6341],
        [0.4901]]) torch.Size([7, 1])
tensor([[0.4963],
        [0.7682],
        [0.0885],
        [0.1320],
        [0.3074],
        [0.6341],
        [0.4901]]) torch.Size([7, 1])
tensor([[0.4963],
        [0.7682],
        [0.0885],
        [0.1320],
        [0.3074],
        [0.6341],
        [0.4901]]) torch.Size([7, 1])


5. Speaking of random seeds, we saw how to set it with `torch.manual_seed()` but is there a GPU equivalent? (hint: you'll need to look into the documentation for `torch.cuda` for this one)

In [4]:
# Q.5
device = "cuda" if torch.cuda.is_available() else "cpu"
print(device)

torch.cuda.manual_seed(RANDOM_SEED)

cuda


6. Create two random tensors of shape (2, 3) and send them both to the GPU (you'll need access to a GPU for this). Set torch.manual_seed(1234) when creating the tensors (this doesn't have to be the GPU random seed). The output should be something like:
Device: cuda
(tensor([[0.0290, 0.4019, 0.2598],
         [0.3666, 0.0583, 0.7006]], device='cuda:0'),
 tensor([[0.0518, 0.4681, 0.6738],
         [0.3315, 0.7837, 0.5631]], device='cuda:0'))

In [5]:
# Q.6
RANDOM_SEED = 1234

torch.manual_seed(RANDOM_SEED)
tensor_A = torch.rand(2, 3)

torch.manual_seed(RANDOM_SEED)
tensor_B = torch.rand(2, 3)

device = "cuda" if torch.cuda.is_available() else "cpu"

tensor_A = tensor_A.to(device)
tensor_B = tensor_B.to(device)

print(tensor_A, tensor_A.device)
print(tensor_B, tensor_B.device)

tensor([[0.0290, 0.4019, 0.2598],
        [0.3666, 0.0583, 0.7006]], device='cuda:0') cuda:0
tensor([[0.0290, 0.4019, 0.2598],
        [0.3666, 0.0583, 0.7006]], device='cuda:0') cuda:0


7. Perform a matrix multiplication on the tensors you created in 6 (again, you may have to adjust the shapes of one of the tensors).
The output should look like:

(tensor([[0.3647, 0.4709],
         [0.5184, 0.5617]], device='cuda:0'), torch.Size([2, 2]))

In [6]:
# Q.7
result_tensor = torch.matmul(tensor_A, tensor_B.mT)    # Transpose
print(result_tensor, result_tensor.shape)

tensor([[0.2299, 0.2161],
        [0.2161, 0.6287]], device='cuda:0') torch.Size([2, 2])


8. Find the maximum and minimum values of the output of 7.

In [7]:
max = result_tensor.max()
min = result_tensor.min()

print(f"Max: {max}")
print(f"Min: {min}")

Max: 0.628727912902832
Min: 0.21611443161964417


9. Find the maximum and minimum index values of the output of 7.


In [8]:
max_index = result_tensor.argmax()
min_index = result_tensor.argmin()

print(f"Index of max element: {max_index}")
print(f"Index of min element: {min_index}")

Index of max element: 3
Index of min element: 1


10. Make a random tensor with shape (1, 1, 1, 10) and then create a new tensor with all the 1 dimensions removed to be left with a tensor of shape (10). Set the seed to 7 when you create it and print out the first tensor and it's shape as well as the second tensor and it's shape.
The output should look like:

tensor([[[[0.5349, 0.1988, 0.6592, 0.6569, 0.2328, 0.4251, 0.2071, 0.6297,
           0.3653, 0.8513]]]]) torch.Size([1, 1, 1, 10])
tensor([0.5349, 0.1988, 0.6592, 0.6569, 0.2328, 0.4251, 0.2071, 0.6297, 0.3653,
        0.8513]) torch.Size([10])

In [9]:
RANDOM_SEED = 7

# First tensor
tensor_A = torch.rand(1, 1, 1, 10)

# Second tensor
torch.manual_seed(RANDOM_SEED)
tensor_B = tensor_A.squeeze()

print(tensor_A, tensor_A.shape)
print(tensor_B, tensor_B.shape)

tensor([[[[0.0518, 0.4681, 0.6738, 0.3315, 0.7837, 0.5631, 0.7749, 0.8208,
           0.2793, 0.6817]]]]) torch.Size([1, 1, 1, 10])
tensor([0.0518, 0.4681, 0.6738, 0.3315, 0.7837, 0.5631, 0.7749, 0.8208, 0.2793,
        0.6817]) torch.Size([10])
